Added some new compiler flag options, enabled and tested dipole field calculation and ability to control output precision

richard-evans opened the pull request:
Merged with latest develop branch of the code. Another big change coming but this will make the final integration easier.

odarbelaeze commented:
@richard-evans can you summarise the changes specific to CUDA for me to check, and maybe outline the error you're getting at runtime. I'm not able to run the code till next week, since we're reallocating our servers.

richard-evans commented:
Yes, added support for selecting DP and SP at compile time, implemented dipole calculation and confirmed it works, and added sync call for config outputs.

richard-evans commented:
Runtime error for large systems with dia matrix is an internal error: 
_terminate called after throwing an instance of 'cusp::format_conversion_exception'
what():  dia_matrix fill-in would exceed maximum tolerance_ apparently due to number of zero's stored in the matrix. Hybrid may actually be the optimal choice - need to check performance
